In [2]:
from sqlalchemy import create_engine, select, Column, Integer, String, Date, ForeignKey, PrimaryKeyConstraint
from sqlalchemy.orm import sessionmaker, relationship
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import create_engine, MetaData, Table

db_string = "postgresql://wbauer_adb:adb2020@pgsql-196447.vipserv.org:5432/wbauer_adb"
db = create_engine(db_string)

metadata = MetaData()

### <span style='color:Yellow'> 1. Calculate the average cost of renting out all your movies.  </span>

In [3]:
stmt = '''
SELECT AVG(amount)
FROM payment;
'''

result = db.execute(stmt).fetchall()
result

[(Decimal('4.2006056453822965'),)]

### <span style='color:red'> Answer: </span> 4.2006056453822965

### <span style='color:Yellow'> 2. Calculate and display the number of videos in all categories.  </span>

In [4]:
stmt = '''
SELECT c.name, COUNT(*)
FROM film AS f1
 INNER JOIN film_category AS f2
  ON f1.film_id = f2.film_id
   INNER JOIN category AS c
    ON f2.category_id = c.category_id
GROUP BY c.name;
'''

result = db.execute(stmt).fetchall()
result

[('Horror', 56),
 ('Comedy', 58),
 ('Sci-Fi', 61),
 ('Drama', 62),
 ('Foreign', 73),
 ('Classics', 57),
 ('Games', 61),
 ('New', 63),
 ('Travel', 57),
 ('Music', 51),
 ('Action', 64),
 ('Documentary', 68),
 ('Sports', 74),
 ('Animation', 66),
 ('Family', 69),
 ('Children', 60)]

### <span style='color:Yellow'> 3. View the count of all customers grouped by country.  </span>

In [5]:
stmt = '''
SELECT c3.country, COUNT(*)
FROM customer AS c1
 INNER JOIN address AS a
  USING (address_id)
   INNER JOIN city AS c2
    USING (city_id)
     INNER JOIN country AS c3
      USING (country_id)
GROUP BY c3.country
ORDER BY c3.country;
'''

result = db.execute(stmt).fetchall()
result

[('Afghanistan', 1),
 ('Algeria', 3),
 ('American Samoa', 1),
 ('Angola', 2),
 ('Anguilla', 1),
 ('Argentina', 13),
 ('Armenia', 1),
 ('Austria', 3),
 ('Azerbaijan', 2),
 ('Bahrain', 1),
 ('Bangladesh', 3),
 ('Belarus', 2),
 ('Bolivia', 2),
 ('Brazil', 28),
 ('Brunei', 1),
 ('Bulgaria', 2),
 ('Cambodia', 2),
 ('Cameroon', 2),
 ('Canada', 5),
 ('Chad', 1),
 ('Chile', 3),
 ('China', 53),
 ('Colombia', 6),
 ('Congo, The Democratic Republic of the', 2),
 ('Czech Republic', 1),
 ('Dominican Republic', 3),
 ('Ecuador', 3),
 ('Egypt', 6),
 ('Estonia', 1),
 ('Ethiopia', 1),
 ('Faroe Islands', 1),
 ('Finland', 1),
 ('France', 4),
 ('French Guiana', 1),
 ('French Polynesia', 2),
 ('Gambia', 1),
 ('Germany', 7),
 ('Greece', 2),
 ('Greenland', 1),
 ('Holy See (Vatican City State)', 1),
 ('Hong Kong', 1),
 ('Hungary', 1),
 ('India', 60),
 ('Indonesia', 14),
 ('Iran', 8),
 ('Iraq', 1),
 ('Israel', 4),
 ('Italy', 7),
 ('Japan', 31),
 ('Kazakstan', 2),
 ('Kenya', 2),
 ('Kuwait', 1),
 ('Latvia', 2),
 (

### <span style='color:Yellow'> 4. Display information about a store that has more than 100 customers and less than 300 customers.  </span>

In [6]:
stmt = '''
SELECT s.store_id, COUNT(*)
FROM customer AS c
 INNER JOIN store AS s
  USING (store_id)
GROUP BY s.store_id
HAVING COUNT(*) BETWEEN 100 AND 300
'''

result = db.execute(stmt).fetchall()
result

[(2, 273)]

### <span style='color:Yellow'> 5. Select all customers who watched movies for more than 200 hours.  </span>

In [27]:
stmt = '''
SELECT c.customer_id, c.first_name, c.last_name, SUM(f.length)
FROM customer AS c
 INNER JOIN rental AS r
  USING (customer_id)
   INNER JOIN inventory AS i
    USING (inventory_id)
     INNER JOIN film AS f
      USING (film_id)
GROUP BY c.customer_id
HAVING SUM(f.length) > 12000;
'''

result = db.execute(stmt).fetchall()
result

[]

### <span style='color:red'> Answer: </span> None

### <span style='color:Yellow'> 6. Calculate the average value of a movie rented.  </span>

In [30]:
stmt = '''
SELECT AVG(p.amount)
FROM rental AS r
 INNER JOIN payment AS p
  USING (rental_id)
'''

result = db.execute(stmt).fetchall()
result

[(Decimal('4.2006056453822965'),)]

### <span style='color:red'> Answer: </span> 4.2006056453822965

### <span style='color:Yellow'> 7. Calculate the average value of the video length in all categories.  </span>

In [32]:
stmt = '''
SELECT c.name, AVG(f1.length)
FROM film AS f1
 INNER JOIN film_category AS f2
  ON f1.film_id = f2.film_id
   INNER JOIN category AS c
    ON f2.category_id = c.category_id
GROUP BY c.name;
'''

result = db.execute(stmt).fetchall()
result

[('Horror', Decimal('112.4821428571428571')),
 ('Comedy', Decimal('115.8275862068965517')),
 ('Sci-Fi', Decimal('108.1967213114754098')),
 ('Drama', Decimal('120.8387096774193548')),
 ('Foreign', Decimal('121.6986301369863014')),
 ('Classics', Decimal('111.6666666666666667')),
 ('Games', Decimal('127.8360655737704918')),
 ('New', Decimal('111.1269841269841270')),
 ('Travel', Decimal('113.3157894736842105')),
 ('Music', Decimal('113.6470588235294118')),
 ('Action', Decimal('111.6093750000000000')),
 ('Documentary', Decimal('108.7500000000000000')),
 ('Sports', Decimal('128.2027027027027027')),
 ('Animation', Decimal('111.0151515151515152')),
 ('Family', Decimal('114.7826086956521739')),
 ('Children', Decimal('109.8000000000000000'))]

### <span style='color:Yellow'> 8. Find longest movie titles in all categories.  </span>

In [43]:
stmt = '''
SELECT c.name, MAX(LENGTH(f1.title))
FROM film AS f1
 INNER JOIN film_category AS f2
  ON f1.film_id = f2.film_id
   INNER JOIN category AS c
    ON f2.category_id = c.category_id
GROUP BY c.name;
'''

result = db.execute(stmt).fetchall()
result

[('Horror', 27),
 ('Comedy', 23),
 ('Sci-Fi', 20),
 ('Drama', 22),
 ('Foreign', 20),
 ('Classics', 23),
 ('Games', 21),
 ('New', 21),
 ('Travel', 23),
 ('Music', 22),
 ('Action', 23),
 ('Documentary', 22),
 ('Sports', 25),
 ('Animation', 22),
 ('Family', 22),
 ('Children', 20)]

### <span style='color:Yellow'> 9. Find longest movie in all categories. </span>

In [52]:
stmt = '''
SELECT c.name, MAX(f1.length)
FROM film AS f1
 INNER JOIN film_category AS f2
  ON f1.film_id = f2.film_id
   INNER JOIN category AS c
    ON f2.category_id = c.category_id
GROUP BY c.name;
'''

result = db.execute(stmt).fetchall()
result

[('Horror', 181),
 ('Comedy', 185),
 ('Sci-Fi', 185),
 ('Drama', 181),
 ('Foreign', 184),
 ('Classics', 184),
 ('Games', 185),
 ('New', 183),
 ('Travel', 185),
 ('Music', 185),
 ('Action', 185),
 ('Documentary', 183),
 ('Sports', 184),
 ('Animation', 185),
 ('Family', 184),
 ('Children', 178)]